### PR TITLE
Erase temporary buffer in EVP_PKEY_get_bn_param()

### PR DESCRIPTION
**Pull request by iamnotacake**
Function EVP_PKEY_get_bn_param() uses temporary buffer (on stack or heap allocated) to store serialized bignum, but after deserializing it into BIGNUM*, the buffer is not erased and may contain sensitive data.

This change makes sure the buffer is erased if it was successfully filled before. Unfortunately, it does not distinguish between public and private key components, and will always erase the buffer.

Fixes https://github.com/openssl/openssl/issues/20626

### CODE REVIEW

File: crypto/evp/p_lib.c [hunk 1]
```diff
@@ -2176,7 +2176,15 @@ int EVP_PKEY_get_bn_param(const EVP_PKEY *pkey, const char *key_name,
         goto err;
     ret = OSSL_PARAM_get_BN(params, bn);
 err:
-    OPENSSL_free(buf);
+    if (buf) {
```
Comment (slontis) [inline crypto/evp/p_lib.c:2179]: if (buf != NULL

File: crypto/evp/p_lib.c [hunk 2]
```diff
@@ -2176,7 +2176,15 @@ int EVP_PKEY_get_bn_param(const EVP_PKEY *pkey, const char *key_name,
...
+        if (OSSL_PARAM_modified(params)) {
```
Comment (slontis) [inline crypto/evp/p_lib.c:2180]: for this inner if where the body of the if and else blocks are single lines, you can remove the {}

**Review by slontis — COMMENTED** [review]
I don't know if we have any way of testing the malloc branch currently since it requires more than 16384 bits of data.

**Comment by iamnotacake** [comment]
> I don't know if we have any way of testing the malloc branch currently since it requires more than 16384 bits of data.

Well, one thing I could try is generate a really large RSA key, with `n` being slightly bigger than 2048 bytes. Hope it's supported and my computer won't die. Then, I can add a test that contains my key as hardcoded bignum arrays, import it with `EVP_PKEY_fromdata()`, then export a large bignum `n` using `EVP_PKET_get_bn_param()`. Using hardcoded key is not an issue in this case and I guess import should work much faster than generation, so no problem in adding it to openssl test suite.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
I think it is fine as is.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master, 3.1, and 3.0 branches. Thank you for your contribution.

**Pull request closed without merging**